python/cris_import.py

```diff
@@ -0,0 +1,257 @@
+        return data["data"]
+    except KeyError:
+        raise ValueError(data)
+
+
+def get_file_meta(filename):
+    """
+    Returns:
+        schema_year, extract_id, table_name
+    """
+    filename_parts = filename.split("_")
+    return filename_parts[1], filename_parts[2], filename_parts[3]
+
+
+def get_files_todo():
+    files_todo = []
+    for filename in os.listdir(FILE_DIR):
+        if not filename.endswith(".csv"):
+            continue
+        schema_year, extract_id, table_name = get_file_meta(filename)
+
+        # ignore all tables except these
+        if table_name not in ("crash", "unit", "person", "primaryperson"):
+            continue
+
+        file = {
+            "schema_year": schema_year,
+            "table_name": table_name,
+            "path": os.path.join(FILE_DIR, filename),
+            "extract_id": extract_id,
+        }
+        files_todo.append(file)
+    return files_todo
+
+
+def load_csv(filename):
+    with open(filename, "r") as fin:
+        reader = csv.DictReader(fin)
+        return [row for row in reader]
+
+
+def lower_case_keys(list_of_dicts):
+    return [{key.lower(): value for key, value in row.items()} for row in list_of_dicts]
+
+
+def remove_unsupported_columns(rows, table_name):
+    col_names_to_keep = [
+        col["name"]
+        for col in COLUMNS
+        if col["table_name"] == table_name and col["source"] == "cris"
+    ]
+    return [
+        {key: val for key, val in row.items() if key in col_names_to_keep}
+        for row in rows
+    ]
+
+
+def handle_empty_strings(rows):
+    for row in rows:
+        for key, val in row.items():
+            if val == "":
+                row[key] = None
+    return rows
+
+
+def handle_crash_date_time(rows):
+    """Combine crash_date and crash_time and format as ISO string.
+
+    The same ISO timestamp is stored as `crash_date` and `crash_time`
+    """
+    tz = pytz.timezone("US/Central")
+    for row in rows:
+        # parse a string that looks like '12/22/2023'
+        crash_date = row["crash_date"]
+        month, day, year = crash_date.split("/")
+        # parse a string that looks like '12:19 PM'
+        crash_time = row["crash_time"]
+        hour_minute, am_pm = crash_time.split(" ")
+        hour, minute = hour_minute.split(":")
+        if am_pm.lower() == "pm":
+            hour = int(hour) + 12 if hour != "12" else int(hour)
+        else:
+            hour = int(hour) if hour != "12" else 0
+        # create a tz-aware instance of this datetime
+        crash_date_dt = tz.localize(
+            datetime(int(year), int(month), int(day), hour=hour, minute=int(minute))
+        )
+        # save the ISO string with tz offset
+        crash_date_iso = crash_date_dt.isoformat()
+        row["crash_date"] = crash_date_iso
+        row["crash_time"] = crash_date_iso
+
+
+def chunks(lst, n):
+    """Yield successive n-sized chunks from lst."""
+    for i in range(0, len(lst), n):
+        yield lst[i : i + n]
```

Comment:
❤️ the generator. Your python is great to learn from in terms of patterns and elegant methods.

Reply:
oh this 100% copypasta from an SO post that i come back to all the time 😅